provider/pkg/schema/gen.go

```diff
@@ -1210,37 +1212,70 @@ func (ctx *context) propertyTypeSpec(parentName string, propSchema *jsschema.Sch
 		}, nil
 	}
 
+	if propSchema.Items != nil && len(propSchema.Items.Schemas) > 0 {
+		elementType, err := ctx.propertyTypeSpec(parentName+"Item", propSchema.Items.Schemas[0])
+		if err != nil {
+			return nil, err
+		}
+		return &pschema.TypeSpec{
+			Type:  "array",
+			Items: elementType,
+		}, nil
+	}
+
+	if propSchema.Items != nil {
+		elementType, err := ctx.propertyTypeSpec(parentName+"Item", propSchema.Items.Schemas[0])
+		if err != nil {
+			return nil, err
+		}
+		return &pschema.TypeSpec{
+			Type:  "array",
+			Items: elementType,
+		}, nil
+	}
+
 	// All other types.
-	if len(propSchema.Type) > 0 {
-		switch propSchema.Type[0] {
-		case jsschema.IntegerType:
-			return &pschema.TypeSpec{Type: "integer"}, nil
-		case jsschema.StringType:
-			return &pschema.TypeSpec{Type: "string"}, nil
-		case jsschema.BooleanType:
-			return &pschema.TypeSpec{Type: "boolean"}, nil
-		case jsschema.NumberType:
-			return &pschema.TypeSpec{Type: "number"}, nil
-		case jsschema.ObjectType:
-			return &pschema.TypeSpec{Ref: "pulumi.json#/Any"}, nil
-		case jsschema.ArrayType:
-			elementType, err := ctx.propertyTypeSpec(parentName+"Item", propSchema.Items.Schemas[0])
-			if err != nil {
-				return nil, err
-			}
-			return &pschema.TypeSpec{
-				Type:  "array",
-				Items: elementType,
-			}, nil
-		default:
+	if len(propSchema.Type) == 1 {
+		typeSpec := parseJsonType(propSchema.Type[0])
+		return &typeSpec, nil
+	}
+
+	if len(propSchema.Type) > 1 {
```

Comment:
I think, to avoid the problem with `type: ["string", "array"]` mentioned above, this check should actually come right after `NormalizeTypes` and we should recurse on `propertySpec` with a schema that's simplified to have a single element for type.

Reply:
Yes, I think you're right that this would be a better design. I've tried to keep the change fairly conservative so far as all changes seem to have fairly wide-reaching changes. Will have a spike on if this is possible to do well. It gets a little more tricky when we can also have OneOfs too .. but perhaps we could safely assume that it can't be multi-typed if it's a one-of.

Reply:
Alas, I've tried moving this up but this check is a less accurate version of the object and array checks meaning that we loose a lot of fidelity. I think this whole area of code needs a big rethink to address these fragile bits.

Reply:
Ok, thanks for giving it a look anyway. I'm fine with just leaving a comment and some tests cases that acknowledge the limitations.